### DISCUSSION TITLE
Difference between Lunatic and Bastion

### DISCUSSION THREAD

**Discussion by zmilan** [lunatic-solutions/lunatic#27] [General] [answered]
Hi,

I'm very new to both of these tools and both look very much awesome to me. But I'm not able to understand deeply what is difference between them except that the Lunatic target is wasm while [Bastion](https://docs.rs/bastion/0.4.3/bastion/) stays in Rust. Could anyone help me to understand this better? Thanks in advance.

**Comment by bkolobara** [comment, marked as answer]
Hi @zmilan!

The biggest difference to me is that Bastion's processes are just [Rust Futures](https://doc.rust-lang.org/std/future/trait.Future.html). This makes them extremely lightweight and fits perfectly into the existing Rust eco-system.

But you can't do some of the stuff that Lunatic enables:

* Limit per process memory/cpu usage.
* Disallow a specific process from reading from filesystem or opening network connections.

From the perspective of a process in Lunatic, it owns the whole memory. You can never read/write directly to the memory of other processes (not even through unsafe pointer dereferencing).

**Reply by zmilan** [reply]
Hi @bkolobara thanks for the explanation. This is exactly the type of information I was looking for. I'm looking forward to fining some good use case and start playing with Lunatic :)
Thanks for awesome work :)

**Reply by aschrijver** [reply]
I didn't know about Bastion, but in 2023 it doesn't look like the project is still actively maintained.
